### ISSUE TITLE
How to expand model.ini files in a more generic or why .ini files?

### ISSUE THREAD

**Issue by goanpeca** [lektor/lektor#485]
I have been talking with @nixjdm a bit about why lektor does some things the way it does, so I want to ask a question regarding how we have custom types and how this handled internally.

First, why do we use of .ini files to define models? I believe the answer is, because this way users do not really need to know/understand python. Is this the reason @mitsuhiko ?

Second, lektor use a custom flat-file based system for defining the content of the site based on the models that are defined. This keeps things simple and human readable which I think is great. Now, it seems we will no need to keep defining custom types and properties for these (such as max, min length, regex validation and so on) and we could avoid duplication or reinventing the wheel if instead we used python libraries that already do this at great length (schematics, marshmallow, sqlalchemy). On the later option we could potentially

- have all the types that something like SQLAlchemy provides, and have the ini file just be a front to keep things simple for non python users?
- Could we also use this integration to eventually build a real DB and populate the content there so maybe we can perform more complicated queries
- or even eventually ease with migration when a model changes something?

Would this be something worth exploring? Creating custom types with regex validation, length validation etc feels like reinventing the wheel with so many other libraries already doing this.

Thoughts @mitsuhiko, @singingwolfboy, @nixjdm?

**Comment by torian257x** [comment]
I agree that these ini files are tedious. A solution that gives more programmatical freedom would very very welcomed. Just a simple example: **have a date field with a prefilled default value of today.**
finally: people end up using python functions all over their ini files. Even for a simple select etc.
